source/extensions/certificate_providers/certificate_provider.h

```diff
@@ -55,14 +56,27 @@ class CertificateProvider {
   virtual const std::string& getCACertificate(absl::string_view cert_name) const PURE;
 
   /*
-   * Get TLS identity certificates directly from provider
+   * Get TLS certpair directly from provider, return identity cert and key for handshake
+   * or ca cert and key for issuer.
    */
-  virtual std::list<Certpair> getIdentityCertificates(absl::string_view cert_name) PURE;
+  virtual std::list<Certpair> getCertpair(absl::string_view cert_name) PURE;
 
   /*
    * Generate TLS identity certificate dynamically during TLS handshake
    */
-  virtual Certpair* generateTlsCertificate(absl::string_view server_name) PURE;
+  virtual Certpair* generateIdentityCertificate(const SSL_CLIENT_HELLO* ssl_client_hello) PURE;
+
+  /*
+   * CertificateSubsriptionCallbacks
+   */
+  virtual void onCertpairUpdated(absl::string_view cert_name, Certpair certpair) PURE;
+  virtual void onCACertUpdated(absl::string_view cert_name, const std::string cert) PURE;
+  virtual void onUpatedFailed() PURE;
+
+  /*
+   * Add subsription on one cert/certpair
+   */
+  virtual void addSubsription(CertificateSubscriptionPtr subscription, std::string cert_name) PURE;
```

Comment:
How do we cancel a subscription when the config changes and we no longer need to subscribe to a cert name?

Reply:
I don't know why we need to unsubscribe to a cert name. Who will do that? I think when an certificate provider is constructed, it's subscriptions are initiated as well at the same time and the subscriptions keep watching the certs update and save back to certificate provider through the subscription callbacks. Actually we even don't need this addSubscription function, we can let each certificate provider impl keeps them as private object.

Reply:
this is outdated, agree with "CertificateProvider is itself the thing that will be subscribing to the external resource"